src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,140 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_INDEX = 0;
+    private static final int INITIAL_CAPACITY = 10;
+    private Object[] defaultArray = new Object[INITIAL_CAPACITY];
+    private int sizeCounter = 0;
+    private int arrayLength = 10;
+
     @Override
     public void add(T value) {
-
+        if (sizeCounter >= arrayLength) {
+            arrayLength = grow();
+            defaultArray = increaseArray(defaultArray);
+        }
+        defaultArray[sizeCounter++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        rangeCheckerForAdding(index);
+        sizeCounter++;
+        if (sizeCounter >= arrayLength) {
+            arrayLength = grow();
+        }
+        defaultArray = increaseArray(defaultArray,value, index);
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (sizeCounter + list.size() > arrayLength) {
+            arrayLength = grow();
+        }
+        defaultArray = increaseArray(list);
+    }
+
+    private Object[] increaseArray(Object[] array) {
+        Object[] temp = new Object[arrayLength];
+        System.arraycopy(array, DEFAULT_INDEX, temp, DEFAULT_INDEX, sizeCounter);
+        return temp;
+    }
+
+    private Object[] increaseArray(Object[] array, T value, int index) {
+        Object[] temp = new Object[arrayLength];
+        System.arraycopy(array,DEFAULT_INDEX, temp, DEFAULT_INDEX, index);
+        System.arraycopy(array, index, temp, index + 1, sizeCounter - index);
+        temp[index] = value;
+        return temp;
```

Comment:
You can simplify your solution. Let's copy the array only one time. You can use next algorithm:
````
Objects[] data = values;
values = new Object[size * 3 / 2];
System.arraycopy(data, 0, values, 0, size);
````